internal/command/project_application_saml_test.go

```diff
@@ -0,0 +1,754 @@
+					MetadataURL: "http://localhost:" + port + path,
+					State:       domain.AppStateActive,
+				},
+			},
+		},
+		{
+			name: "create oidc app metadataURL, http error",
+			fields: fields{
+				eventstore: eventstoreExpect(
+					t,
+					expectFilter(
+						eventFromEventPusher(
+							project.NewProjectAddedEvent(context.Background(),
+								&project.NewAggregate("project1", "org1").Aggregate,
+								"project", true, true, true,
+								domain.PrivateLabelingSettingUnspecified),
+						),
+					),
+				),
+				idGenerator: id_mock.NewIDGeneratorExpectIDs(t),
+				httpStatus:  http.StatusNotFound,
+			},
+			args: args{
+				ctx: context.Background(),
+				samlApp: &domain.SAMLApp{
+					ObjectRoot: models.ObjectRoot{
+						AggregateID: "project1",
+					},
+					AppName:     "app",
+					EntityID:    "https://test.com/saml/metadata",
+					Metadata:    []byte("<?xml version=\"1.0\"?>\n<md:EntityDescriptor xmlns:md=\"urn:oasis:names:tc:SAML:2.0:metadata\"\n                     validUntil=\"2022-08-26T14:08:16Z\"\n                     cacheDuration=\"PT604800S\"\n                     entityID=\"https://test.com/saml/metadata\">\n    <md:SPSSODescriptor AuthnRequestsSigned=\"false\" WantAssertionsSigned=\"false\" protocolSupportEnumeration=\"urn:oasis:names:tc:SAML:2.0:protocol\">\n        <md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat>\n        <md:AssertionConsumerService Binding=\"urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST\"\n                                     Location=\"https://test.com/saml/acs\"\n                                     index=\"1\" />\n        \n    </md:SPSSODescriptor>\n</md:EntityDescriptor>"),
+					MetadataURL: "http://localhost:" + port + path,
+				},
+				resourceOwner: "org1",
+			},
+			res: res{
+				err: errors.IsErrorInvalidArgument,
+			},
+		},
+	}
+	metadata := []byte("")
+	httpStatus := 0
+	http.HandleFunc(path, func(w http.ResponseWriter, req *http.Request) {
+		if httpStatus != 0 {
+			http.Error(w, "error", httpStatus)
+			return
+		}
+		if err := xml.Write(w, metadata); err != nil {
+			t.Error("ReadMetadataFromURL() failed to write metadata")
+			return
+		}
+	})
+	go func() {
+		if err := http.ListenAndServe(":"+port, nil); err != nil {
+			t.Error("ReadMetadataFromURL() failed to ListenAndServe")
+			return
+		}
+	}()
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.args.samlApp.MetadataURL != "" {
+				metadata = tt.args.samlApp.Metadata
+				httpStatus = tt.fields.httpStatus
+				tt.args.samlApp.Metadata = []byte("")
+			}
+
+			r := &Commands{
+				eventstore:  tt.fields.eventstore,
+				idGenerator: tt.fields.idGenerator,
+			}
+
+			got, err := r.AddSAMLApplication(tt.args.ctx, tt.args.samlApp, tt.args.resourceOwner)
```

Comment:
i see several issues with these tests (actual http server, global vars which are changed for individual tests, mixing of args and fields, ...)

I suggest changing the `xml.ReadMetadataFromURL` function to receive an http.Client rather than using the default client (this will also enable to handle proxies and other settings)
the commands struct in zitadel can then store such a client and pass it
it would improve reproducibility also a lot in tests as you could just pass a test client (as field individually in each test), where for example the http.RoundTripper interface is implemented (so no actual call is needed)
further I'd create a var for the xmlMetadata (note you could use []byte(``) to handle multilines there) and use it in args, res and events
and please then only pass metadata in args where necessary 

same obviously also applies to the `ChangeSAML` tests

let me know if you have questions or need some example (if tested it a bit locally)